SpatialGDK/Source/SpatialGDK/Private/Interop/Connection/SpatialEventTracer.cpp

```diff
@@ -0,0 +1,273 @@
+#else
+	using UnrealProperty = UProperty;
+#endif
+
+	for (TFieldIterator<UnrealProperty> It(Struct); It; ++It)
+	{
+		UnrealProperty* Property = *It;
+
+		FString VariableName = Property->GetName();
+		const void* Value = Property->ContainerPtrToValuePtr<uint8>(&EventMessage);
+
+		check(Property->ArrayDim == 1); // Arrays not handled yet
+
+#if ENGINE_MINOR_VERSION >= 25
+		if (FStrProperty* StringProperty = CastField<FStrProperty>(Property))
+#else
+		if (UStrProperty* StringProperty = Cast<UStrProperty>(Property))
+#endif
+		{
+			AddTraceEventStringField(VariableName, StringProperty->GetPropertyValue(Value));
+		}
+#if ENGINE_MINOR_VERSION >= 25
+		else if (FObjectProperty* ObjectProperty = CastField<FObjectProperty>(Property))
+#else
+		else if (UObjectProperty* ObjectProperty = Cast<UObjectProperty>(Property))
+#endif
+		{
+			UObject* Object = ObjectProperty->GetPropertyValue(Value);
+			if (Object)
+			{
+				AddTraceEventStringField(VariableName, Object->GetName());
+				if (AActor* Actor = Cast<AActor>(Object))
+				{
+					FString KeyString = VariableName + TEXT("Position");
+					FString ValueString = Actor->GetTransform().GetTranslation().ToString();
+					AddTraceEventStringField(KeyString, ValueString);
+				}
+			}
+			else
+			{
+				AddTraceEventStringField(VariableName, "Null");
+			}
+		}
+		else // Default
+		{
+			FString StringValue;
+			Property->ExportTextItem(StringValue, Value, NULL, NULL, PPF_None);
+			AddTraceEventStringField(VariableName, StringValue);
+		}
+	}
+
+	// TODO(EventTracer): implement and call AddTargetObjectInfoToEventData
+	// TODO(EventTracer): implement and call AddComponentIdInfoToEventData
+
+	TraceEvent.data = EventData;
+	Trace_EventTracer_AddEvent(EventTracer, &TraceEvent);
+	Trace_EventData_Destroy(EventData);
+
+	return CurrentSpanId;
+}
+
+bool SpatialEventTracer::IsEnabled() const
+{
+	return bEnabled; // Trace_EventTracer_IsEnabled(EventTracer);
+}
+
+void SpatialEventTracer::Enable(const FString& FileName)
+{
+	Trace_EventTracer_Parameters parameters = {};
+	EventTracerGuard.EventTracer = this;
+	parameters.callback = &SpatialEventTracer::TraceCallback;
+	EventTracer = Trace_EventTracer_Create(&parameters);
+	Trace_EventTracer_Enable(EventTracer);
+	bEnabled = true;
+
+	UE_LOG(LogSpatialEventTracer, Log, TEXT("Spatial event tracing enabled."));
+
+	// Open a local file
+	const FString FolderPath = FPaths::Combine(FPaths::ProjectSavedDir(), TEXT("EventTracing"));
+	const FString FullFileName = FString::Printf(TEXT("EventTrace_%s_%s.trace"), *FileName, *FDateTime::Now().ToString());
+	const FString FilePath = FPaths::Combine(FolderPath, FullFileName);
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
+	if (PlatformFile.CreateDirectoryTree(*FolderPath))
+	{
+		UE_LOG(LogSpatialEventTracer, Log, TEXT("Capturing trace to %s."), *FilePath);
+		Stream.Reset(Io_CreateFileStream(TCHAR_TO_ANSI(*FilePath), Io_OpenMode::IO_OPEN_MODE_WRITE));
```

Comment:
This function needs to lock the mutex also to ensure proper memory ordering for the worker thread

Reply:
I think? Not sure if Trace_EventTracer_Enable(EventTracer) kicks off that thread or waits for connection to init

Reply:
Think the thread is created after this code runs actually ignore this